### PR TITLE
feat!: add type field to eik config schema

### PR DESCRIPTION
**Pull request by digitalsadhu**
Adds type field
Changes validation method to no longer mutate subject being validated
Adds validation to config class constructor and setters

BREAKING CHANGE: valid eik config object is now required when instantiating EikConfig class

### CODE REVIEW

File: test/classes/eik-config.test.js [hunk 1]
```diff
@@ -5,77 +5,156 @@ const EikConfig = require('../../lib/classes/eik-config');
...
+    });
+    t.deepEqual(config.map, ['http://map']);
 
-    config = new EikConfig({ 'import-map': ['gouda', 'mozzarella'] });
-    t.deepEqual(config.map, ['gouda', 'mozzarella']);
+    config = new EikConfig({
+        name: 'pizza',
+        server: 'http://server',
+        files: { '/': 'pizza' },
+        version: '0.0.0',
```
Comment (gingermusketeer) [inline test/classes/eik-config.test.js:32]: Perhaps it is worth extracting out a VALID_CONFIG object that can be reused to make the tests less noisy?

**Comment by gingermusketeer** [comment]
Should the commit be marked as a breaking change so that the version gets bumped as expected?

I think this will break https://github.com/eik-lib/common/blob/master/lib/helpers/get-defaults.js#L18 which the CLI relies on. Likely that there aren't any tests for that. Might be good to test this against the CLI locally to be sure that I haven't forgotten about other cases.

File: lib/schemas/validate.js [hunk 2]
```diff
@@ -26,57 +27,61 @@ const eikJSON = createValidator(eikJSONSchema, {
...
+        }
+        if (!pkvalid.validForOldPackages) {
+            errors.push({
+                keyword: 'validForOldPackages',
+                dataPath: '.name',
+                schemaPath: '',
+                params: [],
+                message: 'should be valid for old packages',
+            });
+        }
```
Comment (trygve-lie) [inline lib/schemas/validate.js:51]: Do we need to cater for "old" packages at this stage? The amount of projects using this is probably at a level where we can manually update them in an hour or so.

File: test/classes/eik-config.test.js (hunk 1 above)
Comment (digitalsadhu) [reply test/classes/eik-config.test.js:32]: good idea

**Comment by digitalsadhu** [comment]
@gingermusketeer the commit is marked as breaking. Using feat!: in the commit message and/or using BREAKING CHANGE: in the commit body tell semantic release to bump major. Is that what you mean?

File: lib/schemas/validate.js (hunk 2 above)
Comment (digitalsadhu) [reply lib/schemas/validate.js:51]: I'm squinting at this code (it's only in this PR due to formatting changes) and I think I get what's going on. Seems like we should be able to remove the old validation bit.

**Review by trygve-lie — APPROVED** [review]
lgtm :+1:

**Review by gingermusketeer — COMMENTED** [review]
Might be good to check that an exception is raised if the config is not valid when read from disk or before it is saved to disk. I think it is important that there is a helpful message explaining what is wrong.

**Comment by digitalsadhu** [comment]
@gingermusketeer 
```
InvalidConfigError: Eik config object was invalid: 'config.findInDirectory operation failed: Invalid eik.json schema: should have required property 'server''
```

and

```
InvalidConfigError: Eik config object was invalid: 'config.persistToDisk operation failed: Invalid eik.json schema: should have required property 'server'
```

**Pull request merged**